cpp/src/structs/utilities.cpp

```diff
@@ -207,73 +213,149 @@ flattened_table flatten_nested_columns(table_view const& input,
   return table_flattener{input, column_order, null_precedence, nullability}();
 }
 
-void superimpose_nulls(bitmask_type const* null_mask,
-                       size_type null_count,
-                       column& input,
-                       rmm::cuda_stream_view stream,
-                       rmm::mr::device_memory_resource* mr)
+namespace {
+
+/**
+ * @brief Dispatcher to conservatively check if the input column needs to be sanitized (i.e., if it
+ * may contain non-empty nulls).
+ *
+ * For performance reason, no non-empty null will actually be checked. Instead, this will perform
+ * conservative checking to see if the input column contains any lists or strings column that is
+ * nulllable. As such, it may return false positive answer.
+ */
+struct may_need_sanitize_dispatch {
+  template <typename T>
+  bool operator()(column_view const& col) const
+  {
+    return std::any_of(col.child_begin(), col.child_end(), [](auto const& child) {
+      return type_dispatcher(child.type(), may_need_sanitize_dispatch{}, child);
+    });
+  }
+};
+
+template <>
+bool may_need_sanitize_dispatch::operator()<cudf::string_view>(column_view const& col) const
+{
+  return col.nullable();
+}
+
+template <>
+bool may_need_sanitize_dispatch::operator()<cudf::list_view>(column_view const& col) const
+{
+  // A lists column may have empty data (i.e., having child column of type EMPTY).
+  // Thus, it needs to be sanitized only if the child column is not empty.
+  return col.nullable() &&
+         col.child(lists_column_view::child_column_index).type().id() != type_id::EMPTY;
```

Comment:
Wow, why is this required? Under what scenario is `typeid::EMPTY` possible?

Reply:
I hit list of empty type in json column creation. Without handling empty type here, json tests will fail.